### PR TITLE
Fix Content-Disposition in media repository

### PR DESCRIPTION
**Pull request by hawkowl**

### CODE REVIEW

File: .coveragerc [hunk 1]
```diff
@@ -0,0 +1,12 @@
+[run]
```
Comment (richvdh) [inline .coveragerc]: sorry, but could you put the coverage stuff in a separate PR?

File: synapse/rest/media/v1/_base.py [hunk 2]
```diff
@@ -197,3 +199,64 @@ def __init__(self, server_name, file_id, url_cache=False,
...
+
+
+def get_filename_from_headers(headers):
+    """
+    Get the filename of the downloaded file by inspecting the
+    Content-Disposition HTTP header.
+
+    Args:
+        headers (twisted.web.http_headers.Headers): The HTTP
+        request headers.
```
Comment (richvdh) [inline synapse/rest/media/v1/_base.py]:
```suggestion
            request headers.
```

File: synapse/rest/media/v1/_base.py [hunk 3]
```diff
@@ -197,3 +199,64 @@ def __init__(self, server_name, file_id, url_cache=False,
...
+
+    Returns:
+        A Unicode string of the filename, or None.
+    """
+    content_disposition = headers.get(b"Content-Disposition", [b''])
+
+    # Decode the Content-Disposition header (cgi.parse_header requires
+    # unicode on Python 3, not bytes) the best we can.
+    try:
+        content_disposition = content_disposition[0].decode('utf8')
```
Comment (richvdh) [inline synapse/rest/media/v1/_base.py]: surely this will fail if there was no `content_disposition` header?

Comment (richvdh) [inline synapse/rest/media/v1/_base.py]: surely we shouldn't decode the entire header as utf-8?

File: synapse/rest/media/v1/_base.py [hunk 4]
```diff
@@ -197,3 +199,64 @@ def __init__(self, server_name, file_id, url_cache=False,
...
+
+    # Unquote the string
+    if PY3:
+        upload_name = urllib.parse.unquote(upload_name)
+    else:
+        # Needs to be bytes on Python 2
+        upload_name = urllib.parse.unquote(upload_name.encode('utf8'))
+
+    try:
+        if isinstance(upload_name, bytes):
```
Comment (richvdh) [inline synapse/rest/media/v1/_base.py]: presumably this only happens in the py2 case? can it be moved into that branch of the `if PY3` condition?

File: tests/rest/media/v1/test_media_storage.py [hunk 5]
```diff
@@ -83,3 +88,150 @@ def test_ensure_media_is_in_local_cache(self):
...
+        headers = channel.headers
+        self.assertEqual(headers.getRawHeaders(b"Content-Type"), [b"image/png"])
+        self.assertEqual(
+            headers.getRawHeaders(b"Content-Disposition"),
+            [b"inline; filename*=utf-8''" + filename + b".png"],
+        )
+
+    def test_disposition_filename_utf8escaped(self):
+        """
+        If the filename is `filename=<utf8 escaped>` then Synapse will correctly
```
Comment (richvdh) [inline tests/rest/media/v1/test_media_storage.py]: what does `<utf8 escaped>` mean? %-encoded? Is that even valid?

**Comment by richvdh** [comment]
I'm assuming this is meant to fix https://github.com/matrix-org/synapse/issues/4160

File: synapse/rest/media/v1/_base.py (hunk 2 above)
Comment (hawkowl) [reply synapse/rest/media/v1/_base.py]: It would be nice if we had a docstring format that tools could autoformat for us...

File: synapse/rest/media/v1/_base.py (hunk 3 above)
Comment (hawkowl) [reply synapse/rest/media/v1/_base.py]: No, because above it does a `.get` with the default argument being a list containing empty bytestring.

Comment (hawkowl) [reply synapse/rest/media/v1/_base.py]: Well, we can't decode it as ASCII (which is the original bug), which means that someone is putting UTF-8 literal bytes into the header (non Synapse implementations?). `cgi.parse` needs `str` on Python 3, as per the comment, we have to decode it as something.

Comment (richvdh) [reply synapse/rest/media/v1/_base.py]: oh, duh.

Comment (richvdh) [reply synapse/rest/media/v1/_base.py]: >  someone is putting UTF-8 literal bytes into the header (non Synapse implementations?)

I think the content disposition comes from clients for uploaded media?

> cgi.parse needs str

sounds like `cgi.parse_header` might be broken then :/. My reading of the RFC is that you ought to split out the params *before* you decode it, because different params might be encoded differently, and you don't know what encoding they are using.

OTOH: fair enough, I can see that this is going to work in the majority of cases, and reinventing `parse_header` is probably more hassle than its worth. If you want to take that approach, can you expand the comment to record some of this?

Comment (richvdh) [reply synapse/rest/media/v1/_base.py]: oh wait you rewrote it already. ugh. Ignore the above; I'll re-review this after dinner.

File: .coveragerc (hunk 1 above)
Comment (richvdh) [reply .coveragerc]: bump.

Comment (erikjohnston) [reply .coveragerc]: (Isn't that https://github.com/matrix-org/synapse/pull/4180?)

Comment (hawkowl) [reply .coveragerc]: yes, #4180 :) (If we merge that first then this squashed merge won't have it, which is why I didn't bother removing it)

**Comment by richvdh** [comment]
pending merge of #4180 then

File: synapse/rest/media/v1/_base.py [hunk 6]
```diff
@@ -197,3 +198,71 @@ def __init__(self, server_name, file_id, url_cache=False,
...
+    Returns:
+        A Unicode string of the filename, or None.
+    """
+    content_disposition = headers.get(b"Content-Disposition", [b''])
+
+    # No header, bail out.
+    if not content_disposition[0]:
+        return
+
+    params = {}
```
Comment (richvdh) [inline synapse/rest/media/v1/_base.py]: I'd find a comment which documents the type of this useful. It seems to be a map from a `unicode` to a `bytes`?

(Edit: I see that you kinda document this below, but I'd find it clearer here. Also the terms "decoded" and "unencoded" are pretty overloaded and unclear here)

File: synapse/rest/media/v1/_base.py [hunk 7]
```diff
@@ -197,3 +198,71 @@ def __init__(self, server_name, file_id, url_cache=False,
...
+    upload_name = None
+
+    # First check if there is a valid UTF-8 filename
+    upload_name_utf8 = params.get("filename*", None)
+    if upload_name_utf8:
+        if upload_name_utf8.lower().startswith(b"utf-8''"):
+            upload_name_utf8 = upload_name_utf8[7:]
+            if PY3:
+                try:
+                    # We have a filename*= section. This MUST be ASCII, and any
```
Comment (richvdh) [inline synapse/rest/media/v1/_base.py]: Half of this comment seems to apply to PY2 as well as PY3, so could it be pulled up? Also "quoted" doesn't mean much to me, even if that's what urllib calls it. Can we call it "%-encoded" or "%-escaped" or something?

File: synapse/rest/media/v1/_base.py [hunk 8]
```diff
@@ -197,3 +198,71 @@ def __init__(self, server_name, file_id, url_cache=False,
...
+                    # UTF-8 bytes are quoted. Once it is decoded, we can then
```
Comment (richvdh) [inline synapse/rest/media/v1/_base.py]: "decoded" is unclear to me. Can we say something like:

> First decode the ascii bytes to a str, then we can %-decode it safely

File: synapse/rest/media/v1/_base.py [hunk 9]
```diff
@@ -197,3 +198,71 @@ def __init__(self, server_name, file_id, url_cache=False,
...
+                    # unquote it strictly.
+                    upload_name = urllib.parse.unquote(
+                        upload_name_utf8.decode('ascii'), errors="strict"
+                    )
+                except UnicodeDecodeError:
+                    # Incorrect UTF-8.
+                    pass
+            else:
+                # On Python 2, we can unquote it directly, and then decode it
```
Comment (richvdh) [inline synapse/rest/media/v1/_base.py]:
```suggestion
                # On Python 2, we can %-decode it directly, and then decode the utf8 bytes to a unicode
```
